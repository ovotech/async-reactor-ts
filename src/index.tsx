import * as React from 'react';

interface ChildrenProps {
  result?: any;
  loading: boolean;
  error?: Error;
}

interface Props {
  loader: () => Promise<any>;
  children: (props: ChildrenProps) => JSX.Element;
}

interface State {
  loading: boolean;
  error?: Error;
  result?: any;
}

export class AsyncReactor extends React.Component<Props, State> {
  state = {
    error: undefined,
    loading: true,
    result: undefined,
  };

  componentDidMount() {
    this.props
      .loader()
      .then(result => this.setState({ loading: false, result }), error => this.setState({ loading: false, error }));
  }

  render() {
    return this.props.children({ result: this.state.result, loading: this.state.loading, error: this.state.error });
  }
}
