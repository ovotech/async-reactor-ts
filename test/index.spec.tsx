import { mount } from 'enzyme';
import * as React from 'react';
import { AsyncReactor } from '../src';

describe('Test AsyncReactor', () => {
  test('AsyncReactor should render loading and component', async () => {
    const asyncCondition = new Promise(resolve => setTimeout(() => resolve('some result'), 150));
    const Home = (
      <AsyncReactor loader={() => asyncCondition}>
        {({ loading, result }) => {
          if (loading) {
            return <div>Loading</div>;
          } else {
            return <div>Home Screen {result}</div>;
          }
        }}
      </AsyncReactor>
    );

    const component = mount(Home);

    expect(component.html()).toMatchSnapshot();

    await new Promise(resolve => setTimeout(() => resolve(), 200));
    expect(component.html()).toMatchSnapshot();
  });

  test('AsyncReactor should handle error', async () => {
    const asyncCondition = new Promise((resolve, reject) => setTimeout(() => reject(new Error('some error')), 150));
    const Home = (
      <AsyncReactor loader={() => asyncCondition}>
        {({ loading, result, error }) => {
          if (loading) {
            return <div>Loading</div>;
          } else if (error) {
            return <div>Error</div>;
          } else {
            return <div>Home Screen {result}</div>;
          }
        }}
      </AsyncReactor>
    );

    const component = mount(Home);

    expect(component.html()).toMatchSnapshot();

    await new Promise(resolve => setTimeout(() => resolve(), 200));
    expect(component.html()).toMatchSnapshot();
  });
});
