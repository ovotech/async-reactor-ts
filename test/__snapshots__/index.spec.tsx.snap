// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test AsyncReactor AsyncReactor should handle error 1`] = `"<div>Loading</div>"`;

exports[`Test AsyncReactor AsyncReactor should handle error 2`] = `"<div>Error</div>"`;

exports[`Test AsyncReactor AsyncReactor should render loading and component 1`] = `"<div>Loading</div>"`;

exports[`Test AsyncReactor AsyncReactor should render loading and component 2`] = `"<div>Home Screen some result</div>"`;
